application/account-management/Tests/Api/Users/UserEndpointsTests.cs

```diff
@@ -74,6 +74,88 @@ public async Task GetUser_WhenInvalidUserId_ShouldReturnBadRequest()
         await EnsureErrorStatusCode(response, HttpStatusCode.BadRequest, $"""Failed to bind parameter "UserId Id" from "{invalidUserId}".""");
     }
     
+    [Fact]
+    public async Task GetUsers_WhenSearchingBasedOnUserEmail_ShouldReturnUser()
+    {
+        // Arrange
+        var searchEmail = string.Concat(DatabaseSeeder.User1ForSearching.Email.Take(4));
+        
+        // Act
+        var response = await TestHttpClient.GetAsync($"/api/users?search={searchEmail}");
+        
+        // Assert
+        EnsureSuccessGetRequest(response);
+        var userResponse = await DeserializeResponse<GetUsersResponseDto>(response);
+        userResponse.Should().NotBeNull();
+        userResponse!.TotalCount.Should().Be(1);
+        userResponse.Users.First().Email.Should().Be(DatabaseSeeder.User1ForSearching.Email);
+    }
+    
+    [Fact]
+    public async Task GetUsers_WhenSearchingBasedOnUserFirstName_ShouldReturnUser()
+    {
+        // Arrange
+        var searchEmail = string.Concat(DatabaseSeeder.User1ForSearching.FirstName!.Take(4));
+        
+        // Act
+        var response = await TestHttpClient.GetAsync($"/api/users?search={searchEmail}");
+        
+        // Assert
+        EnsureSuccessGetRequest(response);
+        var userResponse = await DeserializeResponse<GetUsersResponseDto>(response);
+        userResponse.Should().NotBeNull();
+        userResponse!.TotalCount.Should().Be(1);
+        userResponse.Users.First().FirstName.Should().Be(DatabaseSeeder.User1ForSearching.FirstName);
+    }
+    
+    [Fact]
+    public async Task GetUsers_WhenSearchingBasedOnUserLastName_ShouldReturnUser()
+    {
+        // Arrange
+        var searchEmail = string.Concat(DatabaseSeeder.User1ForSearching.LastName!.Take(4));
```

Comment:
Maybe we should have a test that searched for the "Firstname Lastname"?

Again change `var searchEmail`

Reply:
So you mean that we pass one string together with "FirstName Lastname" and then it returns the user? It's not supported right now.

Reply:
No... it should be possible to search for "John Doe" even though it's two separate fields in the Database (because I concatenate the FirstName and LastName before searching). That's why I suggest a test, as this is truly business logic.

Reply:
You concatenate FirstName and LastName before searching ? Can you explain a bit more about that? Can you show me the code of what you are saying?

I would expect this to happen [here](https://github.com/platformplatform/PlatformPlatform/blob/4c11257297c0e8bcc03443ec00825d290561caf9/application/account-management/Infrastructure/Users/UserRepository.cs#L35), but I think that a name concatenation is going to make the query in memory and not in the database.

Reply:
We actually did not allow searching for full name. I add this logic to the search repository and updated the tests. I double checked that it's actually doing this in SQL (by inspecting the log in Aspire). This is how the query looks:

```
SELECT COUNT(*)
FROM [Users] AS [u]
WHERE [u].[Email] LIKE @__search_0_contains ESCAPE N'\' OR COALESCE([u].[FirstName], N'') + N' ' + COALESCE([u].[LastName], N'') LIKE @__search_0_contains ESCAPE N'\'
```
This is OK, the SQL Server is optimized for this kind of queries.

When checking this I also noticed, that there was always an extra SQL Query to get the `TotalCount`. I've added logic to only make that query if the result of the main query returns more than `PageSize` (if not the `TotalCount` is already known).